src/handlers/applicationFormsHandlers.ts

```diff
@@ -0,0 +1,125 @@
+} from '../database';
+import {
+  isApplicationForm,
+  isApplicationFormArray,
+  isQueryContextWithError,
+} from '../types';
+import { ValidationError } from '../errors';
+import type {
+  Request,
+  Response,
+} from 'express';
+import type { Result } from 'tinypg';
+import type { JSONSchemaType } from 'ajv';
+import type { ApplicationForm } from '../types';
+
+const logger = getLogger(__filename);
+
+const getApplicationForms = (req: Request, res: Response): void => {
+  db.sql('applicationForms.selectAll')
+    .then((applicationFormsQueryResult: Result<ApplicationForm>) => {
+      logger.debug(applicationFormsQueryResult);
+      const { rows: applicationForms } = applicationFormsQueryResult;
+      if (isApplicationFormArray(applicationForms)) {
+        res.status(200)
+          .contentType('application/json')
+          .send(applicationForms);
+      } else {
+        throw new ValidationError(
+          'The database responded with an unexpected format.',
+          isApplicationFormArray.errors ?? [],
+        );
+      }
+    })
+    .catch((error: unknown) => {
+      logger.error(error);
+      res.status(500)
+        .contentType('application/json')
+        .send(error);
+    });
+};
+
+const postApplicantionFormsBodySchema: JSONSchemaType<Omit<ApplicationForm, 'createdAt' | 'id' | 'version'>> = {
+  type: 'object',
+  properties: {
+    opportunityId: {
+      type: 'number',
+    },
+  },
+  required: [
+    'opportunityId',
+  ],
+};
+const isPostApplicantionFormsBody = ajv.compile(postApplicantionFormsBodySchema);
+const postApplicationForms = (
+  req: Request<unknown, unknown, Omit<ApplicationForm, 'createdAt' | 'id' | 'version'>>,
+  res: Response,
+): void => {
+  if (!isPostApplicantionFormsBody(req.body)) {
+    res.status(400)
+      .contentType('application/json')
+      .send({
+        message: 'Invalid request body.',
+        errors: isPostApplicantionFormsBody.errors,
+      });
+    return;
+  }
+
+  db.sql('applicationForms.insertOne', {
+    ...req.body,
+    optedIn: false,
+  })
+    .then((applicationFormsQueryResult: Result<ApplicationForm>) => {
+      logger.debug(applicationFormsQueryResult);
+      const applicationForm = applicationFormsQueryResult.rows[0];
+      if (isApplicationForm(applicationForm)) {
+        res.status(201)
+          .contentType('application/json')
+          .send(applicationForm);
+      } else {
+        throw new ValidationError(
+          'The database responded with an unexpected format.',
+          isApplicationForm.errors ?? [],
+        );
+      }
+    })
+    .catch((error: unknown) => {
+      if (error instanceof TinyPgError
+      && isQueryContextWithError(error.queryContext)) {
+        switch (error.queryContext.error.code) {
+          case PostgressErrorCode.FOREIGN_KEY_VIOLATION:
+            res.status(409)
+              .contentType('application/json')
+              .send({
+                message: 'Unique key constraint violation.',
```

Comment:
The message should mention either foreign key violation and/or what can be done about it (e.g. "please create the opportunity first because there is no opportunity with ID=x").

Reply:
That detail is built into the `errors` object (the error has more specifics) -- I think that might be better, since we don't necessarily know what field triggered the error without parsing the error (or without coupling this message with the current schema in ways we might forget to update later).

Does that info resolve the concern or do you want it in the message too?

Reply:
(Oh -- this actually incorrectly says Unique key constraint whoops!  I misread what you were saying!)

Reply:
I think I've resolved this by updating "Unique key" to "Foreign key".

Related note, as mentioned in another comment thread I have a proposed refactor to this logic block (since it ends up being needed in a bunch of places / endpoints) so I think that will be a good time to revisit exploring more detail in the message.